sql/core/src/main/scala/org/apache/spark/sql/execution/Generate.scala

```diff
@@ -56,7 +56,8 @@ case class Generate(
   val boundGenerator = BindReferences.bindReference(generator, child.output)
 
   override def execute(): RDD[Row] = {
-    if (join) {
+    // #SPARK-6489 do not join when the child has no output
+    if (join && child.output.nonEmpty) {
```

Comment:
We should not be trying to optimize degenerate queries such as this one by putting random hacks into the execution engine.